sql: add metric for statement timeouts

spilchen opened the pull request:
Add sql.statement_timeout.count to track the number of SQL statements that fail due to exceeding the statement timeout.

Epic: none
Release note (ops change): Add sql.statement_timeout.count to track the number of SQL statements that fail due to exceeding the statement timeout.

Closes #142008

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/142078)
<!-- Reviewable:end -->

Dedej-Bergin reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142078)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @angles-n-daemons)

___
*[`-- commits` line 2 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142078#-OK76GEz-7ZgvxqBhvZR:-OK76GF-E2Yk4-oYRTwe:b-whji0c):*
The title should be in lowercase like so: `sql: add metric for statement timeouts`

FYI here is where it mentions this in our commit message doc:
 [https://cockroachlabs.atlassian.net/wiki/spaces/CRDB/pages/73072807/Git+Commit+Messages#Commit-title](https://cockroachlabs.atlassian.net/wiki/spaces/CRDB/pages/73072807/Git+Commit+Messages#Commit-title)

![Screenshot 2025-02-27 at 10.56.25 AM.png](https://files.reviewableusercontent.io/images/reviewable/164547647/onZlkprFRX9_rJyhBT_WGi1U/Screenshot%202025-02-27%20at%2010.56.25%E2%80%AFAM.png)


<!-- Sent from Reviewable.io -->

Dedej-Bergin reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142078#-:-OK7AXF_4-yi5FZhJzLo:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @angles-n-daemons and @spilchen)

___
*[`pkg/sql/metric_test.go` line 109 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142078#-OK79t4X2h3cNvYGGPIg:-OK79t4X2h3cNvYGGPIh:b1mxm67) ([raw file](https://github.com/cockroachdb/cockroach/blob/86ee4aa9ddcd856b4546c7691b0a8b13d0d0eed2/pkg/sql/metric_test.go#L109)):*
> ```Go
> 		{query: "COPY mt.n(num) FROM STDIN", copyCount: 1, expectError: true},
> 		{
> 			query:       "BEGIN; SET LOCAL statement_timeout = '10ms'; SELECT pg_sleep(10)",
> ```

nit: Would setting the sleep to 1 be as sufficient for this test as a sleep of 10? i.e. SELECT pg_sleep(1)


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
nice work!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142078#-:-OK7V26Q0TqGy0E-xHEv:b-j4xmku)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @angles-n-daemons and @spilchen)

___
*[`pkg/sql/conn_executor_exec.go` line 228 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142078#-OK7UKuQ59s9Z9cad-p2:-OK7UKuR2lN3n3DN02u4:b5jzunl) ([raw file](https://github.com/cockroachdb/cockroach/blob/86ee4aa9ddcd856b4546c7691b0a8b13d0d0eed2/pkg/sql/conn_executor_exec.go#L228)):*
> ```Go
> 	if errors.Is(p.errorCause(), sqlerrors.QueryTimeoutError) {
> 		ex.metrics.EngineMetrics.StatementTimeoutCount.Inc(1)
> 	}
> ```

what would you think about adding metrics for other reasons the query may have been cancelled?

the two i'm thinking of are transaction_timeout and the client sending a query cancellation request. they can be checked here with `sqlerrors.TxnTimeoutError` and the `QueryCanceled` pgcode (though we could also add a special marker error for the latter too)

if you agree, i'm fine with this PR as-is, and that can be done separately. (also fine if you want to add it here as well)


<!-- Sent from Reviewable.io -->

spilchen committed:
sql: add metric for statement timeouts

Add sql.statement_timeout.count to track the number of SQL statements
that fail due to exceeding the statement timeout.

Epic: none
Release note (ops change): Add sql.statement_timeout.count to track the
number of SQL statements that fail due to exceeding the statement timeout.

Closes #142008

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142078)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @angles-n-daemons, @Dedej-Bergin, and @rafiss)

___
*[`-- commits` line 2 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142078#-OK76GEz-7ZgvxqBhvZR:-OK7VnqO-MAGtXAqbf3Y:boybanx):*
<details><summary><i>Previously, Dedej-Bergin (Bergin Dedej) wrote…</i></summary><blockquote>

The title should be in lowercase like so: `sql: add metric for statement timeouts`

FYI here is where it mentions this in our commit message doc:
 [https://cockroachlabs.atlassian.net/wiki/spaces/CRDB/pages/73072807/Git+Commit+Messages#Commit-title](https://cockroachlabs.atlassian.net/wiki/spaces/CRDB/pages/73072807/Git+Commit+Messages#Commit-title)

![Screenshot 2025-02-27 at 10.56.25 AM.png](https://files.reviewableusercontent.io/images/reviewable/164547647/onZlkprFRX9_rJyhBT_WGi1U/Screenshot%202025-02-27%20at%2010.56.25%E2%80%AFAM.png)
</blockquote></details>

I wasn't aware of this rule. Thanks!
___
*[`pkg/sql/conn_executor_exec.go` line 228 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142078#-OK7UKuQ59s9Z9cad-p2:-OK7XmCu8Nm5-jYZKlzP:b57t79w) ([raw file](https://github.com/cockroachdb/cockroach/blob/86ee4aa9ddcd856b4546c7691b0a8b13d0d0eed2/pkg/sql/conn_executor_exec.go#L228)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

what would you think about adding metrics for other reasons the query may have been cancelled?

the two i'm thinking of are transaction_timeout and the client sending a query cancellation request. they can be checked here with `sqlerrors.TxnTimeoutError` and the `QueryCanceled` pgcode (though we could also add a special marker error for the latter too)

if you agree, i'm fine with this PR as-is, and that can be done separately. (also fine if you want to add it here as well)
</blockquote></details>

Yeah, this is a good idea. I will handle that in a follow-up PR.
___
*[`pkg/sql/metric_test.go` line 109 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142078#-OK79t4X2h3cNvYGGPIg:-OK7XqtS4XxB0EzuU08D:b74uz4c) ([raw file](https://github.com/cockroachdb/cockroach/blob/86ee4aa9ddcd856b4546c7691b0a8b13d0d0eed2/pkg/sql/metric_test.go#L109)):*
<details><summary><i>Previously, Dedej-Bergin (Bergin Dedej) wrote…</i></summary><blockquote>

nit: Would setting the sleep to 1 be as sufficient for this test as a sleep of 10? i.e. SELECT pg_sleep(1)
</blockquote></details>

It would be. However, part of me thinks it might make the test flaky. It's still likely an extreme case, though. If you're okay with that, I'd prefer to leave it at 10.


<!-- Sent from Reviewable.io -->

Dedej-Bergin commented:
*[`pkg/sql/metric_test.go` line 109 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142078#-OK79t4X2h3cNvYGGPIg:-OK7nOuo70_Lnvnr2IjN:bdc188y) ([raw file](https://github.com/cockroachdb/cockroach/blob/86ee4aa9ddcd856b4546c7691b0a8b13d0d0eed2/pkg/sql/metric_test.go#L109)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

It would be. However, part of me thinks it might make the test flaky. It's still likely an extreme case, though. If you're okay with that, I'd prefer to leave it at 10.
</blockquote></details>

Sure, I'm okay with 10.


<!-- Sent from Reviewable.io -->

spilchen commented:
TFTR!

bors r+
